app/models/problem.rb

```diff
@@ -6,6 +6,7 @@
 class Problem
   include Mongoid::Document
   include Mongoid::Timestamps
+  include Mongoid::Attributes::Dynamic
```

Comment:
@biow0lf any comment on this?

Reply:
@stevecrozz @nijikon if comment this include, test suite fails with:

```
  1) Problem notice hosts cache removing a notice removes string from #hosts
     Failure/Error: doc.unset(field)
     
     Mongoid::Errors::UnknownAttribute:
     
       message:
         Attempted to set a value for 'messages.460873b7409daabaaa2ed7bece77e52e' which is not allowed on the model Problem.
       summary:
         Problem#messages.460873b7409daabaaa2ed7bece77e52e= was called but there is no 'messages.460873b7409daabaaa2ed7bece77e52e' field defined in the model, and Mongoid::Attributes::Dynamic is not included. This error is also raised instead of NoMethodError if the unknown attribute is passed to any method that accepts an attributes hash, such as #attributes=.
       resolution:
         Define the field 'messages.460873b7409daabaaa2ed7bece77e52e' in Problem, or include Mongoid::Attributes::Dynamic in Problem if you intend to store values in fields that are not explicitly defined.
     # ./app/models/problem.rb:159:in `block (2 levels) in uncache_notice'
     # ./app/models/problem.rb:152:in `each'
     # ./app/models/problem.rb:152:in `block in uncache_notice'
     # ./app/models/problem.rb:142:in `uncache_notice'
     # ./app/models/notice.rb:132:in `problem_recache'
     # ./spec/models/problem_spec.rb:382:in `block (4 levels) in <top (required)>'
     # ./spec/models/problem_spec.rb:381:in `block (3 levels) in <top (required)>'
```

I am NOT mongodb guru. I never use mongo db. I just don't understand how to fix this.

Including `Mongoid::Attributes::Dynamic` make test suite pass.

And, looks like this is OK. https://jira.mongodb.org/browse/MONGOID-3875

Reply:
OK, I will have a look at this next week. I don't want all your hard work to go to waste.

Reply:
I did some digging into this and I think I found the issue and a possible solution. 

Problem#uncache_notice

```
  def uncache_notice(notice)
    last_notice = notices.last

    atomically do |doc|
      doc.set(
        'environment'    => last_notice.environment_name,
        'error_class'    => last_notice.error_class,
        'last_notice_at' => last_notice.created_at,
        'message'        => last_notice.message,
        'where'          => last_notice.where,
        'notices_count'  => notices_count.to_i > 1 ? notices_count - 1 : 0
      )

      CACHED_NOTICE_ATTRIBUTES.each do |k, v|
        digest = Digest::MD5.hexdigest(notice.send(v))
        field = "#{k}.#{digest}"

        if (doc[k].try(:[], digest).try(:[], :count)).to_i > 1
          doc.inc("#{field}.count" => -1)
        else
          doc.unset(field)
        end
      end
    end
  end
```

`doc.unset(field)` in mongoid >= 7.x interprets this as setting and attribute and since there is no attribute called "#{key}.#{digest}" it fails. Previous version of mongoid appear to just update the hash that #{key} points to. So, looks like we just need to remove the cached digest from the hash using a different method. This would do it:

```
if (doc[k].try(:[], digest).try(:[], :count)).to_i > 1
  doc.inc("#{field}.count" => -1)
else
  h = doc[k]
  h.delete(digest)
  doc.set("#{k}": h)
end
```

With this update and removing `include Mongoid::Attributes::Dynamic` from the entire test suit passes.

Reply:
thanks a ton, @sahglie . so should we move the party over to https://github.com/errbit/errbit/pull/1546 and close this PR?